src/terminal/adapter/FontBuffer.cpp

```diff
@@ -0,0 +1,603 @@
+    const auto textClippingMask = ~(0xFFFF >> _textWidth);
+
+    // If the text is given an explicit width, we check to what extent the
+    // content is offset from center. Knowing that information will enable the
+    // renderer to scale the font more symmetrically.
+    _textCenteringHint = _declaredWidth ? _fullWidth - (_declaredWidth + _textOffset * 2) : 0;
+
+    // Initially the characters are written to the buffer assuming the maximum
+    // cell height, but now that we know the true height, we need to pack the
+    // buffer data so that each character occupies the exact number of scanlines
+    // that are required.
+    for (auto srcLine = 0u, dstLine = 0u; srcLine < _buffer.size(); srcLine++)
+    {
+        if ((srcLine % MAX_HEIGHT) < _fullHeight)
+        {
+            auto characterScanline = _buffer.at(srcLine);
+            characterScanline &= textClippingMask;
+            characterScanline >>= _textOffset;
+            _buffer.at(dstLine++) = characterScanline;
+        }
+    }
+}
+
+void FontBuffer::_fillUnusedCharacters()
+{
+    // Every character in the buffer that hasn't been uploaded will be replaced
+    // with an error glyph (a reverse question mark). This includes every
+    // character prior to the start char, or after the last char.
+    const auto errorPattern = _generateErrorGlyph();
+    for (auto ch = 0u; ch < MAX_CHARS; ch++)
+    {
+        if (ch < _startChar || ch > _lastChar)
+        {
+            auto charBuffer = std::next(_buffer.begin(), ch * _fullHeight);
+            std::copy_n(errorPattern.begin(), _fullHeight, charBuffer);
+        }
+    }
+}
+
+auto FontBuffer::_generateErrorGlyph() -> std::array<uint16_t, MAX_HEIGHT>
+{
+    // We start with a bit pattern for a reverse question mark covering the
+    // maximum font resolution that we might need.
+    constexpr std::array<uint16_t, MAX_HEIGHT> inputBitPattern = {
+        0b000000000000000,
+        0b000000000000000,
+        0b000000000000000,
+        0b000000000000000,
+        0b000000000000000,
+        0b000000000000000,
+        0b001111111110000,
+        0b011111111111000,
+        0b111000000011100,
+        0b111000000011100,
+        0b111000000000000,
+        0b111000000000000,
+        0b111100000000000,
+        0b011111000000000,
+        0b000011110000000,
+        0b000001110000000,
+        0b000001110000000,
+        0b000001110000000,
+        0b000001110000000,
+        0b000000000000000,
+        0b000001110000000,
+        0b000001110000000,
+        0b000001110000000,
+    };
+
+    // Then for each possible width and height, we have hard-coded bit masks
+    // indicating a range of columns and rows to select from the base bitmap
+    // to produce a scaled down version of reasonable quality.
+    constexpr std::array<uint32_t, MAX_WIDTH + 1> widthMasks = {
+        // clang-format off
+        0, 1, 3, 32771, 8457, 9481, 9545, 9673, 42441, 26061,
+        58829, 28141, 60909, 63453, 63485, 65533, 65535

```

Comment:
This is some mind blowing wizardry honestly.
How did you generate the constants? The bit patterns don't appear to me to be arbitrary.

Reply:
This was a very manual process. I started with the full size image, and then one by one I would pick which bit columns to eliminate to make it narrower, while still retaining a reasonable shape. Then I did the same thing vertically, picking which rows to eliminate to make the character shorter.

This was the only way I could get the character to look reasonable at all sizes. Simple scaling it wouldn't work, because at larger sizes you tend to need a lot of padding, while the smaller sizes have almost no padding at all.